[IMP] tests: check is resource file used

samueljlieber opened the pull request:
This PR adds a test to `make review` to check if a resource file is used in it's `.rst` file, when running checks on a resource folder.

![check-is-resource-used](https://github.com/user-attachments/assets/15927190-0626-4a69-8f76-ba05e8cabc7c)

For each resource file in the resource folder, check if the file is referenced in an image or figure directive in the .rst file (as long as the .rst file has the same name as the resource folder).

task-2801043

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/documentation/pull/12010.png)](https://mergebot.odoo.com/odoo/documentation/pull/12010)

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+                file,
+                0,
+                "resouce file not used",
+                'is-resource-file-used',
```
```suggestion
                'resource-file-referenced',
```

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+            resource_folder,
+            0,
+            "resource folder does not match rst file name",
+            'is-resource-file-used',
```
```suggestion
            'resource-folder-match',
```

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
     yield 0, "the file does not have an extension"
+
+@sphinxlint.checker('')
+def check_is_resource_file_used(file, options=None):
```
I'd move this check function above `check_file_extensions` to keep resource file checkers together.

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+
+@sphinxlint.checker('')
+def check_is_resource_file_used(file, options=None):
+    """ Check each resource file is used in it's .rst file. """
```
```suggestion
    """ Check that resource files are referenced in at least one RST file. """
```

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+@sphinxlint.checker('')
+def check_is_resource_file_used(file, options=None):
+    """ Check each resource file is used in it's .rst file. """
+    resource_file = os.path.basename(file)
+    resource_folder = os.path.dirname(file)
+    rst_file = resource_folder + '.rst'
+    image_pattern = re.compile(r'(\.\. (image|figure)::\s*)([^\s]+)')
+    if os.path.exists(rst_file):
+        rst_file_contents = open(rst_file)
```
Except for the line with the RE pattern, all of this should use `pathlib`'s `Path`, not `os`.

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+    resource_file = os.path.basename(file)
+    resource_folder = os.path.dirname(file)
+    rst_file = resource_folder + '.rst'
+    image_pattern = re.compile(r'(\.\. (image|figure)::\s*)([^\s]+)')
```
This works only for `image` and `figure`, but other admonitions or roles should be supported, like `download`.

AntoineVDV replied:
Actually, I think that you could just grep the filename inside the RST content and skip on the first match.

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+    if os.path.exists(rst_file):
+        rst_file_contents = open(rst_file)
+        used_images = set()
+        for lno, line in enumerate(rst_file_contents):
```
Why use `enumerate` if you don't need `lno`?

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+            log_error(
+                file,
+                0,
+                "resouce file not used",
```
```suggestion
                f"the resource file is not referenced in {rst_file}",
```

AntoineVDV replied:
Pending comment

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -61,3 +63,34 @@ def check_resource_file_name(file_path):
+        log_error(
+            resource_folder,
+            0,
+            "resource folder does not match rst file name",
```
```suggestion
            f"the resource folder name doesn't match an RST file name",
```

AntoineVDV commented on tests/main.py:
```diff
@@ -20,6 +20,7 @@
 ADDITIONAL_CHECKERS = [
     checkers.resource_files.check_image_size,
     checkers.resource_files.check_resource_file_name,
+    checkers.resource_files.check_is_resource_file_used,
```
The additional checks are only run in review mode, but this new check should be part of the standard test suite run by Runbot.

AntoineVDV reviewed (changes requested):
Hi @samueljlieber, that looks interesting! Please find my review below.

Ideally, this PR should come with a second commit that fixes the issues found with the new check, so that it can be used on existing content.

I listed your PR in the meta task [2801043](https://www.odoo.com/odoo/project/3835/tasks/2801043) to keep track of the test coverage ;)

samueljlieber commented:
Hi @AntoineVDV! Thank you for your insightful review! In https://github.com/odoo/documentation/commit/27876c56c06998d5cfabe1ebb402a33094b41412 I addressed your suggestions to:
- use pathlib's Path, not os AND grep the filename inside the RST content
  - This greatly simplified the function!
- this new check should be part of the standard test suite run by Runbot
  - I moved this check to run as part of the standard tests
- Ideally, this PR should come with a second commit that fixes the issues found with the new check, so that it can be used on existing content.
  - I made 2 commits to fix the issues exposed by this check:
    1. [[FIX] clean unused resource files](https://github.com/odoo/documentation/pull/12010/commits/1f69939a582745af2e6f950e386e367eeaa68d65): removes all unused resource files flagged by check
    2. [[FIX] resource folder match](https://github.com/odoo/documentation/pull/12010/commits/27876c56c06998d5cfabe1ebb402a33094b41412): corrects all resource folder names that did not match the rst file name

Thanks and I'm looking forward to your thoughts again on these changes! 🙂

AntoineVDV commented on tests/main.py:
```diff
@@ -56,6 +56,7 @@ def run_additional_checks(argv=None):
 - role-without-backticks: Search roles without backticks.
 - trailing-whitespace: Check for trailing whitespaces at end of lines.
 - unbalanced-inline-literals-delimiters: Search for unbalanced inline literals delimiters.
+- resource-file-referenced: Check for unreferenced resource files.
```
No need to list it here as it's part of "all the checkers defined in checkers/* files" mentioned below.

AntoineVDV commented on tests/main.py:
```diff
@@ -85,6 +86,10 @@ def run_additional_checks(argv=None):
         re.compile(rf'\.\.\. {sphinxlint.ALL_DIRECTIVES}::'),
     ):
         parser = argparse.ArgumentParser()
+        _enabled_checkers, args = sphinxlint.parse_args()
+        for path in chain.from_iterable(sphinxlint.walk(path, args.ignore) for path in args.paths):
+            if path.startswith('content') and not path.endswith('.rst'):
+                checkers.resource_files.check_resource_file_referenced(path)
```
This is essentially a copy-paste of `run_additional_checks`. Try this instead to reuse the code but dynamically change which checks should be called:
```diff
diff --git a/tests/main.py b/tests/main.py
index 8d7de7760..870aef42d 100644
--- a/tests/main.py
+++ b/tests/main.py
@@ -18,8 +18,7 @@ CUSTOM_RST_DIRECTIVES = [
 ]
 
 ADDITIONAL_CHECKERS = [
-    checkers.resource_files.check_image_size,
-    checkers.resource_files.check_resource_file_name,
+    checkers.resource_files.check_resource_file_referenced,
 ]
 
 
@@ -88,6 +87,10 @@ if __name__ == '__main__':
         if os.getenv('REVIEW') == '1':  # Enable checkers for `make review`.
             setattr(sphinxlint.check_line_too_long, 'enabled', True)
             setattr(checkers.rst_style.check_early_line_breaks, 'enabled', True)
-            ADDITIONAL_CHECKERS.extend([checkers.resource_files.check_image_color_depth])
-            run_additional_checks()
+            ADDITIONAL_CHECKERS.extend([
+                checkers.resource_files.check_image_size,
+                checkers.resource_files.check_image_color_depth,
+                checkers.resource_files.check_resource_file_name,
+            ])
+        run_additional_checks()
         sys.exit(sphinxlint.main())
```

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -57,6 +57,27 @@ def check_resource_file_name(file_path):
 
+def check_resource_file_referenced(file, options=None):
+    """ Check that resource files are referenced in at least one RST file. """
+    resource_file = Path(file).name
+    resource_folder = Path(file).parent
+    rst_file = resource_folder.with_suffix('.rst')
+    if rst_file.exists():
+        if resource_file not in rst_file.read_text():
```
```suggestion
    resource_file = Path(file)
    resource_folder = resource_file.parent
    rst_file = resource_folder.with_suffix('.rst')
    if rst_file.exists():
        if resource_file.name not in rst_file.read_text():
```

samueljlieber commented:
Thank you again @AntoineVDV! Addressed your suggestions in https://github.com/odoo/documentation/commit/b9099e98d6e1c65047c1eda4f6139b2fe2c22cab, and resolved your pending comment.

I added the condition `if path.startswith('content')` to run_additional_checks since there isn't a reason to run checks on resource_files outside of this folder (as far as I know).

AntoineVDV commented on tests/checkers/resource_files.py:
```diff
@@ -57,6 +57,27 @@ def check_resource_file_name(file_path):
+            log_error(
+                file,
+                0,
+                "resouce file not used",
+                f"the resource file is not referenced in {rst_file}",
```
This is the other way around:
```suggestion
                f"the resource file is not referenced in {rst_file}",
                resource-file-referenced,
```

AntoineVDV reviewed (approved):
LGTM, thanks!

@robodoo r+ rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.
